src/certificate/commands/DefectsCommand.ts

```diff
@@ -0,0 +1,151 @@
+import { Service } from 'typedi';
+import { DefectRepository } from '../../defect/DefectRepository';
+import { DefectService } from '../../defect/DefectService';
+import { ITestResult } from '../../models';
+import { ICertificatePayload } from '../../models';
+import { ITestType } from '../../models';
+import { CERTIFICATE_DATA, TEST_RESULTS } from '../../models/Enums';
+import { IDefectParent } from '../../models/IDefectParent';
+import { IFlatDefect } from '../../models/IFlatDefect';
+
+@Service()
+export class DefectsCommand {
+	protected type?: CERTIFICATE_DATA;
+
+	protected isWelsh = false;
+
+	constructor(
+		private defectService: DefectService,
+		private defectRepository: DefectRepository
+	) {}
+
+	private certificateIsAnPassOrFail = (): boolean =>
+		this.type === CERTIFICATE_DATA.PASS_DATA || this.type === CERTIFICATE_DATA.FAIL_DATA;
+
+	public initialise(type: CERTIFICATE_DATA, isWelsh = false) {
+		this.type = type;
+		this.isWelsh = isWelsh;
+	}
+
+	public async generate(testResult: ITestResult): Promise<ICertificatePayload> {
+		if (!this.certificateIsAnPassOrFail()) {
+			return {} as ICertificatePayload;
+		}
+
+		const { testTypes } = testResult;
+
+		const result = {} as ICertificatePayload;
+
+		if (testTypes.testResult !== TEST_RESULTS.FAIL) {
+			result.DATA = {
+				...(await this.getPayloadData(testResult, CERTIFICATE_DATA.PASS_DATA)),
+			};
+		}
+
+		if (testTypes.testResult !== TEST_RESULTS.PASS) {
+			result.FAIL_DATA = {
+				...(await this.getPayloadData(testResult, CERTIFICATE_DATA.FAIL_DATA)),
+			};
+		}
+
+		return result;
+	}
+
+	private async getPayloadData(testResult: ITestResult, type: CERTIFICATE_DATA): Promise<any> {
+		let defectListFromApi: IDefectParent[] = [];
+		let flattenedDefects: IFlatDefect[] = [];
+
+		if (this.isWelsh) {
+			defectListFromApi = await this.defectRepository.getDefectTranslations();
+			flattenedDefects = this.defectService.flattenDefectsFromApi(defectListFromApi);
+		}
+
+		const defects = await this.generateDefects(
+			testResult.testTypes,
+			type,
+			testResult.vehicleType,
+			flattenedDefects,
+			this.isWelsh
+		);
+		return defects;
+	}
+
+	/**
+	 * Generates an object containing defects for a given test type and certificate type
+	 * @param testTypes - the source test type for defect generation
+	 * @param type - the certificate type
+	 * @param vehicleType - the vehicle type from the test result
+	 * @param flattenedDefects - the list of flattened defects after being retrieved from the defect service
+	 * @param isWelsh - determines whether the atf in which the test result was conducted resides in Wales
+	 */
+	private generateDefects(
```

Comment:
In the current develop even though this code is near enough identical, it seems to be implicit that it is an async function. But for some reason it no longer seems to be picking that up with this change, so the linting is coming up with `Redundant 'await' for a non-promise type` on line 63 for me.

Reply:
I had this when using eslint. Is your linting running biome?

Reply:
This function is a hot mess tbh, with the mix of by ref / by val params and lack of typings 😬

Reply:
Strange, yes I am using Biome for this repo with ESLint disabled.

Reply:
There is also a `tslint.json` in this repo which is causing issues in my IDE - it might be the right time to strip that out!